R/drive_ls.R

```diff
@@ -26,51 +28,161 @@
 #' }
 #'
 #' @export
-drive_ls <- function(search = NULL, ..., fixed = FALSE, verbose = TRUE){
+drive_ls <- function(path = NULL,
+                     pattern = NULL,
+                     ...,
+                     fixed = FALSE,
+                     verbose = TRUE){
+
+  folder <- NULL
 
-  request <- build_drive_ls(...)
+  if (!is.null(path)){
+    folder <- climb_folders(path = path)
+  }
+
+  request <- build_drive_ls(..., folder = folder)
   response <- make_request(request)
-  process_drive_ls(response = response, search = search, fixed = fixed, verbose = verbose)
+  process_drive_ls(
+    response = response,
+    pattern = pattern,
+    fixed = fixed,
+    verbose = verbose
+  )
+
+}
+
+climb_folders <- function(path = NULL, my_folders = drive_folders()){
+
+  folders <- unlist(strsplit(path, "/"))
+  my_folder_ids <- purrr::map(folders, folder_ids, folder_tbl = my_folders)
+
+  if (!any(my_folder_ids[[1]]$root)){
+    spf("We could not find a folder named '%s' in your 'My Drive' (root) directory.", folders[1])
+  }
 
+  #it is v silly that Google Drive allows this but...
+  if (!sum(my_folder_ids[[1]]$root) == 1){
+    spf("It seems you have more than one folder named '%s' in your 'My Drive' (root) directory.", folders[1])
+  }
+
+  if (length(my_folder_ids) == 1){
+    return(my_folder_ids[[1]]$id[my_folder_ids[[1]]$root == TRUE])
+  }
+  len <- length(my_folder_ids)
+
+  while (nrow(my_folder_ids[[len - 1]]) != 1){
```

Comment:
The problem I'm facing is the iterative nature (not knowing how many to write without a `while`) because if you input `foo/bar/baz` I need to check 2 things, is there a folder named `baz` in `bar` and then if so is that same folder `bar` in `foo`, but if I have `foo/bar/baz/yo` then I need to check 3 things, is `yo` in `baz` and is `baz` in `bar` and is `bar` in `foo`. Other than using a `for` loop (are those better?) I couldn't think of a simpler way (but very open to suggestions!)